File: Leetcode/0787 Cheapest Flights Within K Stops.py
from collections import deque

INF = float('inf')


class Solution:
    def findCheapestPrice(self, n, flights, src, dst, k):
        adj_list = {}
        for flight in flights:
            if flight[0] not in adj_list:
                adj_list[flight[0]] = []
            adj_list[flight[0]].append(flight)

        costs = [INF for _ in range(n)]
        queue = deque([(src, 0)])
        stops = 0
        while stops <= k and queue:
            nodes_in_level = len(queue)
            for _ in range(nodes_in_level):
                node, node_cost = queue.popleft()
                for _, child, edge_cost in adj_list.get(node, []):
                    cost_with_edge = node_cost + edge_cost
                    if cost_with_edge < costs[child]:
                        costs[child] = cost_with_edge
                        queue.append((child, costs[child]))
            stops += 1

        return -1 if costs[dst] is INF else costs[dst]

File: Leetcode/0787 Cheapest Flights Within K Stops.java
class Solution {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        Map<Integer, List<int[]>> adjList = new HashMap<>(n);
        for (var flight : flights) {
            adjList.computeIfAbsent(flight[0], v -> new ArrayList<>()).add(flight);
        }
        int[] cost = new int[n];
        Arrays.fill(cost, Integer.MAX_VALUE);

        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{src, 0});
        int stops = 0;
        while (stops <= k && !queue.isEmpty()) {
            int nodesInLevel = queue.size();
            for (int i = 0; i < nodesInLevel; ++i) {
                var nodeCost = queue.poll();
                for (var edge : adjList.getOrDefault(nodeCost[0], List.of())) {
                    int child = edge[1];
                    int costWithEdge = edge[2] + nodeCost[1];
                    if (costWithEdge < cost[child]) {
                        cost[child] = costWithEdge;
                        queue.add(new int[]{child, cost[child]});
                    }
                }
            }
            ++stops;
        }
        return cost[dst] == Integer.MAX_VALUE ? -1 : cost[dst];
    }
}
